ci2/ci/github.py

```diff
@@ -210,24 +307,37 @@ def __init__(self, branch):
         for pr in self.prs.values():
             await pr.refresh_review_state(gh)
 
-    async def heal(self, batch):
-        # FIXME merge
-        merge_candidate = None
-        for pr in self.prs.values():
-            if pr.review_state == 'approved' and pr.passing is None:
-                merge_candidate = pr.number
-                break
+    async def heal(self, batch_client):
+        if self.healing:
+            self.changed = True
+            return
+
+        self.healing = True
+        self.changed = True
+        while self.changed:
+            self.changed = False
+
+            # FIXME merge
+            merge_candidate = None
+            for pr in self.prs.values():
+                if pr.review_state == 'approved' and pr.build_state is None:
+                    merge_candidate = pr
+                    break
+
+            running_batches = await batch_client.list_batches(
+                complete=False,
+                attributes={
+                    'target_branch': self.branch.short_str()
+                })
 
-        running_batches = await batch.list_batches(
-            complete=False,
-            attributes={
-                'target_branch': self.branch.short_str()
-            })
+            seen_batch_ids = set()
+            for pr in self.prs.values():
+                await pr.heal(batch_client, (merge_candidate is None or pr == merge_candidate), seen_batch_ids)
 
-        seen_batch_ids = set()
-        for number, pr in self.prs.items():
-            await pr.heal(batch, (merge_candidate is None or pr.number == merge_candidate), seen_batch_ids)
+            for batch in running_batches:
+                if batch.id not in seen_batch_ids:
+                    await batch.cancel()
 
-        for batch in running_batches:
-            if batch.id not in seen_batch_ids:
-                await batch.cancel()
+            if merge_candidate and merge_candidate.build_state is not None:
+                self.changed = True
```

Comment:
what's the reasoning for this?

Reply:
If the merge_candidate fails to merge, we want to try again immediately.